Mathlib/RingTheory/Congruence/Hom.lean

```diff
@@ -0,0 +1,500 @@
+  from a smaller to a larger quotient
+
+* `RingCon.quotientKerEquivRangeS`, `RingCon.quotientKerEquivRange`,
+  `RingCon.quotientKerEquivRangeₐ` :
+  the first isomorphism theorem for semirings (using `RingHom.rangeS`),
+  rings (using `RingHom.range`) and algebras (using `AlgHom.range`).
+* `RingCon.comapQuotientEquivRangeS`, `RingCon.comapQuotientEquivRange`,
+  `RingCon.comapQuotientEquivRangeₐ` : the second isomorphism theorem
+  for semirings (using `RingHom.rangeS`), rings (using `RingHom.range`)
+  and algebras (using `AlgHom.range`).
+
+* `RingCon.quotientQuotientEquivQuotient`, `RingCon.quotientQuotientEquivQuotientₐ` :
+  the third isomorphism theorem for semirings (or rings) and algebras
+
+## Tags
+
+congruence, congruence relation, quotient, quotient by congruence relation, ring,
+quotient ring
+-/
+
+
+variable {M : Type*} {N : Type*} {P : Type*}
+
+open Function Setoid
+
+namespace RingCon
+
+section
+
+variable [Mul M] [Add M] [Mul N] [Add N] [Mul P] [Add P]
+  {F} [FunLike F M N] [MulHomClass F M N] [AddHomClass F M N]
+
+/-- The kernel of a ring homomorphism as a ring congruence relation. -/
+def ker (f : F) : RingCon M where
+  toSetoid := Setoid.ker f
+  add' h1 h2 := by
+    dsimp [Setoid.ker, onFun] at *
+    rw [map_add, h1, h2, map_add]
+  mul' h1 h2 := by
+    dsimp [Setoid.ker, onFun] at *
+    rw [map_mul, h1, h2, map_mul]
+
+/-- The definition of the ring congruence relation defined by a ring homomorphism's kernel. -/
+theorem ker_apply (f : F) {x y} : ker f x y ↔ f x = f y :=
+  Iff.rfl
+
+end
+
+section
+
+variable [NonAssocSemiring M] [NonAssocSemiring N] [NonAssocSemiring P]
+  {F} [FunLike F M N] [RingHomClass F M N]
+
+/-- Makes an isomorphism of quotients by two ring congruence
+relations, given that the relations are equal. -/
+protected def congr {c d : RingCon M} (h : c = d) :
+    c.Quotient ≃+* d.Quotient :=
+  { Quotient.congr (Equiv.refl M) <| by apply RingCon.ext_iff.mp h with
+    map_add' x y := by rcases x with ⟨⟩; rcases y with ⟨⟩; rfl
+    map_mul' x y := by rcases x with ⟨⟩; rcases y with ⟨⟩; rfl }
+
+theorem congr_mk {c d : RingCon M} (h : c = d) (a : M) :
+    RingCon.congr h (a : c.Quotient) = (a : d.Quotient) := rfl
+
+theorem comap_eq {c : RingCon M} {f : N →+* M} :
+    c.comap f = ker (c.mk'.comp f) :=
+  ext fun x y => show c _ _ ↔ c.mk' _ = c.mk' _ by rw [← c.eq]; rfl
+
+theorem comap_rel {c : RingCon M} {f : N →+* M} {x y : N} :
+    c.comap f x y ↔ c (f x) (f y) := Iff.rfl
+
+/-- The kernel of the quotient map induced by a ring congruence relation `c` equals `c`. -/
+theorem ker_mk'_eq (c : RingCon M) : ker c.mk' = c :=
+  ext fun _ _ => Quotient.eq''
+
+/-- Given a function `f`, the smallest ring congruence relation containing the binary
+relation on `f`'s image defined by '`x ≈ y` iff the elements of `f⁻¹(x)` are related to
+the elements of `f⁻¹(y)` by a ring congruence relation `c`.' -/
+def mapGen {c : RingCon M} (f : M → N) : RingCon N :=
+  ringConGen <| Relation.Map c f f
```

Comment:
I understand that we have AddCon.mapGen and this is following that approach; but would it make sense to only take `f : M →+* N`? Is that the intended use case?